File: examples/Java/peering3.java
package guide;

import java.util.ArrayList;
import java.util.Random;

import org.zeromq.*;
import org.zeromq.ZMQ.Poller;
import org.zeromq.ZMQ.Socket;

//  Broker peering simulation (part 3)
//  Prototypes the full flow of status and tasks

public class peering3
{

    private static final int    NBR_CLIENTS  = 10;
    private static final int    NBR_WORKERS  = 5;
    // Signals worker is ready
    private static final String WORKER_READY = "\001";

    //  Our own name; in practice this would be configured per node
    private static String self;

    //  This is the client task. It issues a burst of requests and then sleeps
    //  for a few seconds. This simulates sporadic activity; when a number of
    //  clients are active at once, the local workers should be overloaded. The
    //  client uses a REQ socket for requests and also pushes statistics to the
    //  monitor socket:
    private static class client_task extends Thread
    {
        @Override
        public void run()
        {
            try (ZContext ctx = new ZContext()) {
                Socket client = ctx.createSocket(SocketType.REQ);
                client.connect(String.format("ipc://%s-localfe.ipc", self));
                Socket monitor = ctx.createSocket(SocketType.PUSH);
                monitor.connect(String.format("ipc://%s-monitor.ipc", self));
                Random rand = new Random(System.nanoTime());

                Poller poller = ctx.createPoller(1);
                poller.register(client, Poller.POLLIN);

                boolean done = false;
                while (!done) {
                    try {
                        Thread.sleep(rand.nextInt(5) * 1000);
                    }
                    catch (InterruptedException e1) {
                    }
                    int burst = rand.nextInt(15);

                    while (burst > 0) {
                        String taskId = String.format(
                            "%04X", rand.nextInt(10000)
                        );
                        //  Send request, get reply
                        client.send(taskId, 0);

                        //  Wait max ten seconds for a reply, then complain
                        int rc = poller.poll(10 * 1000);
                        if (rc == -1)
                            break; //  Interrupted

                        if (poller.pollin(0)) {
                            String reply = client.recvStr(0);
                            if (reply == null)
                                break; //  Interrupted
                            //  Worker is supposed to answer us with our task id
                            assert (reply.equals(taskId));
                            monitor.send(String.format("%s", reply), 0);
                        }
                        else {
                            monitor.send(
                                String.format(
                                    "E: CLIENT EXIT - lost task %s", taskId
                                ),
                            0);
                            done = true;
                            break;
                        }
                        burst--;
                    }
                }
            }
        }
    }

    //  This is the worker task, which uses a REQ socket to plug into the LRU
    //  router. It's the same stub worker task you've seen in other examples:

    private static class worker_task extends Thread
    {
        @Override
        public void run()
        {
            Random rand = new Random(System.nanoTime());
            try (ZContext ctx = new ZContext()) {
                Socket worker = ctx.createSocket(SocketType.REQ);
                worker.connect(String.format("ipc://%s-localbe.ipc", self));

                //  Tell broker we're ready for work
                ZFrame frame = new ZFrame(WORKER_READY);
                frame.send(worker, 0);

                while (true) {
                    //  Send request, get reply
                    ZMsg msg = ZMsg.recvMsg(worker, 0);
                    if (msg == null)
                        break; //  Interrupted

                    //  Workers are busy for 0/1 seconds
                    try {
                        Thread.sleep(rand.nextInt(2) * 1000);
                    }
                    catch (InterruptedException e) {
                    }

                    msg.send(worker);
                }
            }
        }
    }

    //  The main task begins by setting-up all its sockets. The local frontend
    //  talks to clients, and our local backend talks to workers. The cloud
    //  frontend talks to peer brokers as if they were clients, and the cloud
    //  backend talks to peer brokers as if they were workers. The state
    //  backend publishes regular state messages, and the state frontend
    //  subscribes to all state backends to collect these messages. Finally,
    //  we use a PULL monitor socket to collect printable messages from tasks:
    public static void main(String[] argv)
    {
        //  First argument is this broker's name
        //  Other arguments are our peers' names
        //
        if (argv.length < 1) {
            System.out.println("syntax: peering3 me {you}");
            System.exit(-1);
        }
        self = argv[0];
        System.out.printf("I: preparing broker at %s\n", self);
        Random rand = new Random(System.nanoTime());

        try (ZContext ctx = new ZContext()) {
            //  Prepare local frontend and backend
            Socket localfe = ctx.createSocket(SocketType.ROUTER);
            localfe.bind(String.format("ipc://%s-localfe.ipc", self));
            Socket localbe = ctx.createSocket(SocketType.ROUTER);
            localbe.bind(String.format("ipc://%s-localbe.ipc", self));

            //  Bind cloud frontend to endpoint
            Socket cloudfe = ctx.createSocket(SocketType.ROUTER);
            cloudfe.setIdentity(self.getBytes(ZMQ.CHARSET));
            cloudfe.bind(String.format("ipc://%s-cloud.ipc", self));

            //  Connect cloud backend to all peers
            Socket cloudbe = ctx.createSocket(SocketType.ROUTER);
            cloudbe.setIdentity(self.getBytes(ZMQ.CHARSET));
            int argn;
            for (argn = 1; argn < argv.length; argn++) {
                String peer = argv[argn];
                System.out.printf(
                    "I: connecting to cloud forintend at '%s'\n", peer
                );
                cloudbe.connect(String.format("ipc://%s-cloud.ipc", peer));
            }

            //  Bind state backend to endpoint
            Socket statebe = ctx.createSocket(SocketType.PUB);
            statebe.bind(String.format("ipc://%s-state.ipc", self));

            //  Connect statefe to all peers
            Socket statefe = ctx.createSocket(SocketType.SUB);
            statefe.subscribe(ZMQ.SUBSCRIPTION_ALL);
            for (argn = 1; argn < argv.length; argn++) {
                String peer = argv[argn];
                System.out.printf(
                    "I: connecting to state backend at '%s'\n", peer
                );
                statefe.connect(String.format("ipc://%s-state.ipc", peer));
            }

            //  Prepare monitor socket
            Socket monitor = ctx.createSocket(SocketType.PULL);
            monitor.bind(String.format("ipc://%s-monitor.ipc", self));

            //  Start local workers
            int worker_nbr;
            for (worker_nbr = 0; worker_nbr < NBR_WORKERS; worker_nbr++)
                new worker_task().start();

            //  Start local clients
            int client_nbr;
            for (client_nbr = 0; client_nbr < NBR_CLIENTS; client_nbr++)
                new client_task().start();

            //  Queue of available workers
            int localCapacity = 0;
            int cloudCapacity = 0;
            ArrayList<ZFrame> workers = new ArrayList<ZFrame>();

            //  The main loop has two parts. First we poll workers and our two
            //  service sockets (statefe and monitor), in any case. If we have
            //  no ready workers, there's no point in looking at incoming
            //  requests. These can remain on their internal 0MQ queues:
            Poller primary = ctx.createPoller(4);
            primary.register(localbe, Poller.POLLIN);
            primary.register(cloudbe, Poller.POLLIN);
            primary.register(statefe, Poller.POLLIN);
            primary.register(monitor, Poller.POLLIN);

            Poller secondary = ctx.createPoller(2);
            secondary.register(localfe, Poller.POLLIN);
            secondary.register(cloudfe, Poller.POLLIN);

            while (true) {
                //  First, route any waiting replies from workers

                //  If we have no workers anyhow, wait indefinitely
                int rc = primary.poll(localCapacity > 0 ? 1000 : -1);
                if (rc == -1)
                    break; //  Interrupted

                //  Track if capacity changes during this iteration
                int previous = localCapacity;

                //  Handle reply from local worker
                ZMsg msg = null;
                if (primary.pollin(0)) {
                    msg = ZMsg.recvMsg(localbe);
                    if (msg == null)
                        break; //  Interrupted
                    ZFrame address = msg.unwrap();
                    workers.add(address);
                    localCapacity++;

                    //  If it's READY, don't route the message any further
                    ZFrame frame = msg.getFirst();
                    String frameData = new String(frame.getData(), ZMQ.CHARSET);
                    if (frameData.equals(WORKER_READY)) {
                        msg.destroy();
                        msg = null;
                    }
                }
                //  Or handle reply from peer broker
                else if (primary.pollin(1)) {
                    msg = ZMsg.recvMsg(cloudbe);
                    if (msg == null)
                        break; //  Interrupted
                    //  We don't use peer broker address for anything
                    ZFrame address = msg.unwrap();
                    address.destroy();
                }
                //  Route reply to cloud if it's addressed to a broker
                for (argn = 1; msg != null && argn < argv.length; argn++) {
                    byte[] data = msg.getFirst().getData();
                    if (argv[argn].equals(new String(data, ZMQ.CHARSET))) {
                        msg.send(cloudfe);
                        msg = null;
                    }
                }
                //  Route reply to client if we still need to
                if (msg != null)
                    msg.send(localfe);

                //  If we have input messages on our statefe or monitor sockets
                //  we can process these immediately:

                if (primary.pollin(2)) {
                    String peer = statefe.recvStr();
                    String status = statefe.recvStr();
                    cloudCapacity = Integer.parseInt(status);
                }
                if (primary.pollin(3)) {
                    String status = monitor.recvStr();
                    System.out.println(status);
                }

                //  Now we route as many client requests as we have worker
                //  capacity for. We may reroute requests from our local
                //  frontend, but not from the cloud frontend. We reroute
                //  randomly now, just to test things out. In the next version
                //  we'll do this properly by calculating cloud capacity.

                while (localCapacity + cloudCapacity > 0) {
                    rc = secondary.poll(0);

                    assert (rc >= 0);

                    if (secondary.pollin(0)) {
                        msg = ZMsg.recvMsg(localfe);
                    }
                    else if (localCapacity > 0 && secondary.pollin(1)) {
                        msg = ZMsg.recvMsg(cloudfe);
                    }
                    else break; //  No work, go back to backends

                    if (localCapacity > 0) {
                        ZFrame frame = workers.remove(0);
                        msg.wrap(frame);
                        msg.send(localbe);
                        localCapacity--;

                    }
                    else {
                        //  Route to random broker peer
                        int random_peer = rand.nextInt(argv.length - 1) + 1;
                        msg.push(argv[random_peer]);
                        msg.send(cloudbe);
                    }
                }

                //  We broadcast capacity messages to other peers; to reduce
                //  chatter we do this only if our capacity changed.
                if (localCapacity != previous) {
                    //  We stick our own address onto the envelope
                    statebe.sendMore(self);
                    //  Broadcast new capacity
                    statebe.send(String.format("%d", localCapacity), 0);
                }
            }
            //  When we're done, clean up properly
            while (workers.size() > 0) {
                ZFrame frame = workers.remove(0);
                frame.destroy();
            }
        }
    }
}

File: examples/Go/peering3.go
// Broker peering simulation (part 3)
// Prototypes the full flow of status and tasks
//
// Author: amyangfei <amyangfei@gmail.com>
// Requires: http://github.com/alecthomas/gozmq

package main

import (
	"fmt"
	zmq "github.com/alecthomas/gozmq"
	"math/rand"
	"os"
	"strconv"
	"time"
)

const NBR_CLIENTS = 10
const NBR_WORKERS = 5
const WORKER_READY = "\001"

func client_task(name string, i int) {
	context, _ := zmq.NewContext()
	client, _ := context.NewSocket(zmq.REQ)
	monitor, _ := context.NewSocket(zmq.PUSH)
	defer context.Close()
	defer client.Close()
	defer monitor.Close()
	client.SetIdentity(fmt.Sprintf("Client-%s-%d", name, i))
	client.Connect(fmt.Sprintf("ipc://%s-localfe.ipc", name))
	monitor.Connect(fmt.Sprintf("ipc://%s-monitor.ipc", name))

	for {
		time.Sleep(time.Duration(rand.Intn(5)) * time.Second)
		burst := rand.Intn(15)
		for burst > 0 {
			burst--
			task_id := fmt.Sprintf("%04X", rand.Intn(0x10000))

			// Send request with random hex ID
			client.Send([]byte(task_id), 0)

			// Wait max ten seconds for a reply, then complain
			pollset := zmq.PollItems{
				zmq.PollItem{Socket: client, Events: zmq.POLLIN},
			}
			zmq.Poll(pollset, 10*time.Second)
			if pollset[0].REvents&zmq.POLLIN != 0 {
				reply, err := client.Recv(0)
				if err != nil {
					break
				}
				if string(reply) != task_id {
					panic("Worker is supposed to answer us with our task id")
				}
				monitor.Send(reply, 0)
			} else {
				monitor.Send([]byte(fmt.Sprintf("E: CLIENT EXIT - lost task %s", task_id)), 0)
			}
		}
	}
}

func worker_task(name string, i int) {
	context, _ := zmq.NewContext()
	worker, _ := context.NewSocket(zmq.REQ)
	defer context.Close()
	defer worker.Close()
	worker.SetIdentity(fmt.Sprintf("Worker-%s-%d", name, i))
	worker.Connect(fmt.Sprintf("ipc://%s-localbe.ipc", name))

	// Tell broker we're ready for work
	worker.Send([]byte(WORKER_READY), 0)

	// Process messages as they arrive
	for {
		msg, err := worker.RecvMultipart(0)
		if err != nil {
			break
		}
		// Workers are busy for 0/1 seconds
		time.Sleep(time.Duration(rand.Intn(2)) * time.Second)
		fmt.Printf("Worker-%s-%d done: %s\n", name, i, msg)
		worker.SendMultipart(msg, 0)
	}
}

func main() {
	//  First argument is this broker's name
	//  Other arguments are our peers' names
	if len(os.Args) < 2 {
		fmt.Println("syntax: peering3 me {you}...")
		return
	}
	myself := os.Args[1]
	fmt.Printf("I: preparing broker at %s...\n", myself)
	rand.Seed(time.Now().UnixNano())

	context, _ := zmq.NewContext()
	defer context.Close()

	// Prepare local frontend and backend
	localfe, _ := context.NewSocket(zmq.ROUTER)
	localbe, _ := context.NewSocket(zmq.ROUTER)
	defer localfe.Close()
	defer localbe.Close()
	localfe.Bind(fmt.Sprintf("ipc://%s-localfe.ipc", myself))
	localbe.Bind(fmt.Sprintf("ipc://%s-localbe.ipc", myself))

	// Bind cloud fronted to endpoint
	cloudfe, _ := context.NewSocket(zmq.ROUTER)
	defer cloudfe.Close()
	cloudfe.SetIdentity(myself)
	cloudfe.Bind(fmt.Sprintf("ipc://%s-cloud.ipc", myself))

	// Connect cloud backend to all peers
	cloudbe, _ := context.NewSocket(zmq.ROUTER)
	defer cloudbe.Close()
	cloudbe.SetIdentity(myself)
	for i := 2; i < len(os.Args); i++ {
		peer := os.Args[i]
		fmt.Printf("I: connecting to cloud frontend at '%s'\n", peer)
		cloudbe.Connect(fmt.Sprintf("ipc://%s-cloud.ipc", peer))
	}
	// Bind state backend to endpoint
	statebe, _ := context.NewSocket(zmq.PUB)
	defer statebe.Close()
	bindAddress := fmt.Sprintf("ipc://%s-state.ipc", myself)
	statebe.Bind(bindAddress)

	// Connect state frontend to all peers
	statefe, _ := context.NewSocket(zmq.SUB)
	defer statefe.Close()
	statefe.SetSubscribe("")
	for i := 2; i < len(os.Args); i++ {
		peer := os.Args[i]
		fmt.Printf("I: connecting to state backend at '%s'\n", peer)
		statefe.Connect(fmt.Sprintf("ipc://%s-state.ipc", peer))
	}
	// Prepare monitor socket
	monitor, _ := context.NewSocket(zmq.PULL)
	defer monitor.Close()
	monitor.Bind(fmt.Sprintf("ipc://%s-monitor.ipc", myself))

	// Start local workers
	for i := 0; i < NBR_WORKERS; i++ {
		go worker_task(myself, i)
	}

	// Start local clients
	for i := 0; i < NBR_CLIENTS; i++ {
		go client_task(myself, i)
	}

	// Queue of available workers
	local_capacity := 0
	cloud_capacity := 0
	workers := make([]string, 0)

	pollerbe := zmq.PollItems{
		zmq.PollItem{Socket: localbe, Events: zmq.POLLIN},
		zmq.PollItem{Socket: cloudbe, Events: zmq.POLLIN},
		zmq.PollItem{Socket: statefe, Events: zmq.POLLIN},
		zmq.PollItem{Socket: monitor, Events: zmq.POLLIN},
	}

	for {
		timeout := time.Second
		if len(workers) == 0 {
			timeout = -1
		}
		// If we have no workers anyhow, wait indefinitely
		zmq.Poll(pollerbe, timeout)

		// Track if capacity changes during this iteration
		previous := local_capacity
		var msg [][]byte = nil
		var err error = nil

		if pollerbe[0].REvents&zmq.POLLIN != 0 {
			msg, err = localbe.RecvMultipart(0)
			if err != nil {
				break
			}
			identity, _ := msg[0], msg[1]
			msg = msg[2:]
			workers = append(workers, string(identity))
			local_capacity++

			// If it's READY, don't route the message any further
			if string(msg[len(msg)-1]) == WORKER_READY {
				msg = nil
			}
		} else if pollerbe[1].REvents&zmq.POLLIN != 0 {
			msg, err = cloudbe.RecvMultipart(0)
			if err != nil {
				break
			}
			// We don't use peer broker identity for anything
			msg = msg[2:]
		}

		if msg != nil {
			identity := string(msg[0])
			for i := 2; i < len(os.Args); i++ {
				// Route reply to cloud if it's addressed to a broker
				if identity == os.Args[i] {
					cloudfe.SendMultipart(msg, 0)
					msg = nil
					break
				}
			}
			// Route reply to client if we still need to
			if msg != nil {
				localfe.SendMultipart(msg, 0)
			}
		}

		// Handle capacity updates
		if pollerbe[2].REvents&zmq.POLLIN != 0 {
			msg, _ := statefe.RecvMultipart(0)
			status := msg[1]
			cloud_capacity, _ = strconv.Atoi(string(status))
		}

		// handle monitor message
		if pollerbe[3].REvents&zmq.POLLIN != 0 {
			msg, _ := monitor.Recv(0)
			fmt.Println(string(msg))
		}

		for (local_capacity + cloud_capacity) > 0 {
			secondary := zmq.PollItems{
				zmq.PollItem{Socket: localfe, Events: zmq.POLLIN},
			}
			if local_capacity > 0 {
				secondary = append(secondary, zmq.PollItem{Socket: cloudfe, Events: zmq.POLLIN})
			}
			zmq.Poll(secondary, 0)

			if secondary[0].REvents&zmq.POLLIN != 0 {
				msg, _ = localfe.RecvMultipart(0)
			} else if len(secondary) > 1 && secondary[1].REvents&zmq.POLLIN != 0 {
				msg, _ = cloudfe.RecvMultipart(0)
			} else {
				break
			}

			if local_capacity > 0 {
				var worker string
				worker, workers = workers[0], workers[1:]
				msg = append(msg[:0], append([][]byte{[]byte(worker), []byte("")}, msg[0:]...)...)
				localbe.SendMultipart(msg, 0)
				local_capacity--
			} else {
				// Route to random broker peer
				randPeer := rand.Intn(len(os.Args)-2) + 2
				msg = append(msg[:0], append([][]byte{[]byte(os.Args[randPeer]), []byte("")}, msg[0:]...)...)
				cloudbe.SendMultipart(msg, 0)
			}
		}
		if local_capacity != previous {
			statebe.SendMultipart([][]byte{[]byte(myself), []byte(strconv.Itoa(local_capacity))}, 0)
		}
	}
}
